py3status/modules/mpris.py

```diff
@@ -0,0 +1,385 @@
+
+            _time_ms = self._player.Position
+            time = _get_time_str(_time_ms)
+
+            metadata = self._player.Metadata
+            if len(metadata) > 0:
+                is_stream = 'file://' not in metadata.get('xesam:url')
+                title = metadata.get('xesam:title') or UNKNOWN
+                album = metadata.get('xesam:album') or UNKNOWN
+
+                if metadata.get('xesam:artist') is not None:
+                    artist = metadata.get('xesam:artist')[0]
+                else:
+                    # we assume here that we playing a video and these types of
+                    # media we handle just like streams
+                    is_stream = True
+
+                _length_ms = metadata.get('mpris:length')
+                if _length_ms is not None:
+                    length = _get_time_str(_length_ms)
+            else:
+                # use stream format if no metadata is available
+                is_stream = True
+        except Exception:
+            color = i3s_config['color_bad']
+
+        if is_stream:
+            # delete the file extension
+            title = re.sub(r'\....$', '', title)
+            format = self.format_stream
+        else:
+            format = self.format
+
+        return (format.format(player=player,
+                              state=state,
+                              album=album,
+                              artist=artist,
+                              length=length,
+                              time=time,
+                              title=title), color)
+
+    def _get_control_states(self):
+        control_states = {
+            'pause':    _get_ctr_struct(self._player.CanPause,
+                                        self.icon_pause,
+                                        self._player.Pause),
+            'play':     _get_ctr_struct(self._player.CanPlay,
+                                        self.icon_play,
+                                        self._player.Play),
+            'stop':     _get_ctr_struct(True,
+                                        self.icon_stop,
+                                        self._player.Stop),
+            'next':     _get_ctr_struct(self._player.CanGoNext,
+                                        self.icon_next,
+                                        self._player.Next),
+            'previous': _get_ctr_struct(self._player.CanGoPrevious,
+                                        self.icon_previous,
+                                        self._player.Previous)
+        }
+
+        state = 'pause' if self._player.PlaybackStatus == 'Playing' else 'play'
+        control_states['play_pause'] = control_states[state]
+
+        return control_states
+
+    def _get_response_buttons(self):
+        response = []
+        buttons_order = self.buttons_order.split(',')
+        available = ['pause', 'play', 'play_pause', 'stop', 'next', 'previous']
```

Comment:
isn't this just the keys of `_control_state` dict so doesn't need repeating.

`for button in [e for e in buttons_order if e in self._control_states]:`